bodhi/messages/schemas/base.py

```diff
@@ -0,0 +1,232 @@
+from fedora_messaging.schema_utils import user_avatar_url
+
+
+SCHEMA_URL = 'https://bodhi.fedoraproject.org/message-schemas'
+
+
+class BodhiMessage(message.Message):
+    """A base class for Bodhi messages."""
+
+    @property
+    def app_icon(self) -> str:
+        """
+        Return a URL that points to the application's icon.
+
+        This is used when displaying the message to users.
+
+        Returns:
+            A URL for Bodhi's app icon.
+        """
+        return "https://apps.fedoraproject.org/img/icons/bodhi.png"
+
+    @property
+    def agent(self) -> typing.Union[str, None]:
+        """Return the agent's username for this message.
+
+        Returns:
+            The agent's username, or None if the body has no agent key.
+        """
+        return self.body.get('agent', None)
+
+    @property
+    def agent_avatar(self) -> typing.Union[None, str]:
+        """
+        Return a URL to the avatar of the user who caused the action.
+
+        Returns:
+            The URL to the user's avatar, or None if username is None.
+        """
+        username = self.agent
+        if username is None:
+            return None
+        return user_avatar_url(username)
+
+    @property
+    def usernames(self) -> typing.List[str]:
+        """
+        List of users affected by the action that generated this message.
+
+        Returns:
+            A list of affected usernames.
+        """
+        users = []
+        if self.agent is not None:
+            users.append(self.agent)
+
+        if 'comment' in self.body:
+            text = self.body['comment']['text']
+            mentions = re.findall(r'@\w+', text)
+            for mention in mentions:
+                users.append(mention[1:])
+
+        users = list(set(users))
+        users.sort()
+        return users
+
+    @property
+    def containers(self) -> typing.Iterable[str]:
+        """
+        List of containers affected by the action that generated this message.
+
+        Returns:
+            A list of affected container names.
+        """
+        return []
+
+    @property
+    def modules(self) -> typing.Iterable[str]:
+        """
+        List of modules affected by the action that generated this message.
+
+        Returns:
+            A list of affected module names.
+        """
+        return []
+
+    @property
+    def flatpaks(self) -> typing.Iterable[str]:
+        """
+        List of flatpaks affected by the action that generated this message.
+
+        Returns:
+            A list of affected flatpaks names.
+        """
+        return []
+
+
+class BuildV1(typing.NamedTuple):
+    """
+    A model for referencing a Build.
+
+    Attributes:
+        nvr: The koji id of the build.
+    """
+
+    nvr: str
+
+    @property
+    def package(self) -> str:
+        """Return the name of the package that this build is associated with."""
+        return self.nvr.rsplit('-', 2)[0]
+
+    @staticmethod
+    def schema() -> dict:
```

Comment:
What's the rationale for using a `schema()` method instead of the `body_schema` class attribute that the root `Message` class has?

Reply:
Heh, well I had originally made it a classmethod that I was passing some parameters to so the caller could specify the description, but then decided that was overcomplicated and changed it to be this, but it didn't occur to me to make it consistent with the ```Message``` class (note, this is actually not a ```Message```, it's just a ```NamedTuple```). However, I think it would be nice to be consistent, so I'll change it to use the ```body_schema``` as you suggest.

Reply:
@abompard oh you know what? I'm not sure I can easily make it be body_schema. Since this is a ```NamedTuple```, that would mean that body_schema becomes one of the tuple attributes, which I'm not sure we actually want to do. However, I think I could name the method ```body_schema``` if you like. Maybe I could make it be an ```@property``` as a workaround? It wouldn't use ```self```, but might be a nice way to get a similar 'API'?

Reply:
When I think about it though, I tend to dislike making instance methods that don't actually use self. And actually, we also want to be able to use this without instantiating one. I can make it a ```classmethod``` and ```property``` as a better workaround.

Reply:
Technically the classmethod also doesn't use it's ```cls``` parameter, but I guess that's OK.

Reply:
One more thing that might not fit nicely is that this isn't actually a ```body_schema``` anyway - it's not a complete schema, it's just a snippet of a schema. So it wouldn't be the schema for any actual body. A better name might just be ```schema_snippet```?

Reply:
Ah you're right, I thought it was a `Message` subclass. Then the staticmethod is fine, both names are OK (`schema` and `schema_snippet`).

Reply:
@abompard After all the above thoughts, I'm kind of inclined to just leave it as is. It really is a staticmethod, but you can't @property a static method, and it also really isn't a body_schema, because it's not a complete schema, so I think the variation on the name is sensible.

What are your thoughts?

Reply:
Yeah it's fine